### PR TITLE
Simplify and improve typing of `watch` utility

### PR DESCRIPTION
**Pull request by robertknight**
Simplify and improve type safety of the `watch` utility that we use in various services to react to store changes. This utility used to be able to accept either a single function to compute the watched value or an array of functions. It now just accepts a single function, which makes it simpler to specify how the types of the other arguments relate to the watched value. There is also a new argument that explicitly specifies the function used to check whether the current and previous watched values are equal or not. If omitted (set to `null`) it defaults to a strict equality check.

In the process of updating consumers I found that the `FeaturesService` was re-sending feature flags to the host frame every time the guest frames changed, which is pointless. It now only re-sends feature flags to the host frame when the flags themselves change. In future we'll have to figure out how to communicate flags from the sidebar to the guest frames.

### CODE REVIEW

**Comment by lyzadanger** [comment]
@robertknight A question before I dive in on the argument ordering for `watch`. I'm wondering if it might be possible to swap the positions of the `callback` and `compare` arguments (and make `compare` an optional argument) for better ergonomics. Most of the time the `compare` fn is not provided, and the `null` params clutter up the invocations. I have a sense, though, I might be overlooking something (like, maybe this API is modeled after a standard pattern). Thoughts?

File: src/sidebar/util/watch.js [hunk 1]
```diff
@@ -1,81 +1,68 @@
...
+ *     // Extract some data of interest from the store.
+ *     () => store.getValue(),
  *
- *     // Update previous value.
- *     prevValue = new value;
- *   }
- * });
- * ```
+ *     // Use strict comparison of values
+ *     null,
```
Comment (lyzadanger) [inline src/sidebar/util/watch.js:16]: Looks like the argument ordering needs updating in these comments since you swapped `callback` and `compare` ordering.

File: src/sidebar/services/frame-sync.js [hunk 2]
```diff
@@ -134,86 +135,99 @@ export class FrameSyncService {
...
+            prevPublicAnns = publicAnns;
           }
         }
       }
+    };
+
+    watch(
+      this._store.subscribe,
+      () =>
+        /** @type {const} */ ([
```
Comment (lyzadanger) [inline src/sidebar/services/frame-sync.js:224]: Eh?

File: src/sidebar/services/frame-sync.js [hunk 3]
```diff
@@ -134,86 +135,99 @@ export class FrameSyncService {
...
+          this._store.allAnnotations(),
+          this._store.frames(),
+        ]),
+      ([annotations, frames], [prevAnnotations]) =>
```
Comment (lyzadanger) [inline src/sidebar/services/frame-sync.js:228]: Why is the second argument `[prevAnnotations]` instead of `[prevAnnotations, prevFrames]`? Or is it just that `prevFrames` is not used?

File: src/sidebar/util/test/watch-test.js [hunk 4]
```diff
@@ -47,45 +47,42 @@ describe('sidebar/util/watch', () => {
...
+      const getValue = () => [store.getState().a];
 
-      const newEmptyObject = () => ({});
-      watch(store.subscribe, newEmptyObject, callback);
+      watch(store.subscribe, getValue, callback);
 
-      store.dispatch({ type: 'INCREMENT_A' });
-      store.dispatch({ type: 'INCREMENT_A' });
+      store.dispatch({ type: 'INCREMENT_B' });
+      store.dispatch({ type: 'INCREMENT_B' });
```
Comment (lyzadanger) [inline src/sidebar/util/test/watch-test.js:58]: Just curious why this change?

**Review by lyzadanger — APPROVED** [review]
Looks like some comments are out of date since your last changes, but otherwise LGTM!

File: src/sidebar/services/frame-sync.js (hunk 2 above)
Comment (robertknight) [reply src/sidebar/services/frame-sync.js:224]: This is a [const assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions), which has the effect of making TypeScript infer the type as a fixed-sized tuple (here `[Annotation[], Frame[]]`) rather than a more generic array (`Array<Annotation|Frame>`). This results in the `annotations`, `frames` and `prevAnnotations` arguments below all having the expected types.

File: src/sidebar/services/frame-sync.js (hunk 3 above)
Comment (robertknight) [reply src/sidebar/services/frame-sync.js:228]: Indeed, `prevFrames` is not used.

File: src/sidebar/util/watch.js (hunk 1 above)
Comment (robertknight) [reply src/sidebar/util/watch.js:16]: Ah, thanks for catching. Will fix.

File: src/sidebar/util/test/watch-test.js (hunk 4 above)
Comment (robertknight) [reply src/sidebar/util/test/watch-test.js:58]: This is to create a state where the result of `getValue` changes each time if a strict-equality check is used (which we are), but not if a deep equality check were used (which we are not). Here the store states looks like:

1. `{ a: 0, b: 0 }`
2. `{  a: 0, b: 1 }`
3.  `{ a: 0, b: 2 }`

And the sequence of `getValue` results is `[0]`, `[0]`, `[0]`.
